### ISSUE TITLE
Refactor Python/C++ implementation of sliding window ops (Conv/Maxpool ops) to ease up model bring up

### ISSUE THREAD

**Issue by tt-nshanker** [tenstorrent/tt-metal#7083]
Two issues with current implementation -
1) For sliding window ops, we create halo config tensors during op initialization. To construct these halo config tensors, we need the full tensor shape (batch, height, width). As I/O tensors are not available until runtime, we have to manually calculate and track the intermediate tensor shapes during model initialization within __init__ of each module. 
2) Sharding configs are calculated during conv op initialization. This makes conv op init API complicated with a lot of arguments. The sharded ops before and after conv rely on the configs calculated by conv op for resharding at runtime.

From Javelin CNN standup meeting:
Possible solutions and TODO -
1) Break up conv op init functionality. Expose calculation of shard configs as a helper function that can be called either during init or during forward when we have tensor shape.
2) Simplify conv op initialization - only create weight and bias tensors. Defer halo config tensors creation for later. Conv op init api would look the same as Pytorch Conv2d op init API. Not required to pass in tensor shape.
3) Create config tensors during forward pass. Happens in the first iteration where the tensors will be cached. To ensure that all config tensors are stacked on top of l1, we can pre-allocate that space (user configurable) and force TT-metal allocator to allocate in that space. We can tie the config tensors to tt-metal program object which can own them and be responsible for destroying the config tensors at the end of the program.

**Comment by tt-nshanker** [comment]
@nsmithtt @AleksKnezevic @jvasilje @davorchap

**Comment by nsmithtt** [comment]
One gotcha I just realized is that we need a means for tracking original [n, h, w] dims that ideally are propagated via the tensor.  Perhaps we can introduce the concept of `logical_shape` vs `physical_shape` or something.

**Comment by tt-nshanker** [comment]
> One gotcha I just realized is that we need a means for tracking original [n, h, w] dims that ideally are propagated via the tensor. Perhaps we can introduce the concept of `logical_shape` vs `physical_shape` or something.

Yes. But as a first step, we could track tensor height/width during runtime forward calls.

**Comment by tt-nshanker** [comment]
Update > Proposing JIT Conv 

1. Minimal work during model initialization/preprocessing. `pre_process_model` handles folding of conv weights for unity stride and batch norm folding into conv weights. All done on host. Weight/bias tensors are NOT moved to device here.
2. Very minimal runtime code in python. Most of the runtime code (invoked by python forward call) will be moved to C++.
3. Config tensors for reader pattern will be created/allocated on device within `create_program` and tied to program object. With program caching enabled, this will happen only once and config tensors will be cached and reused by conv ops with same parameters.
4. Weights and bias will be converted to desired layout on host and moved to device during runtime just before running conv operation.

**Comment by mywoodstock** [comment]
followups:
https://github.com/tenstorrent/tt-metal/issues/8744
https://github.com/tenstorrent/tt-metal/issues/8745
